float32[] data_x
float32[] data_y
float32[] data_flag